### PR TITLE
POC - Convert RIF layout annotation processors into maven plugins

### PR DESCRIPTION
**Pull request by brianburton**
**JIRA Ticket:**
[BFD-123456](https://jira.cms.gov/browse/BFD-123456)

**User Story or Bug Summary:**

Intellij IDEA is unable to execute the annotation processors in the `bfd-model-codegen` module using its internal maven build process.  This creates a number of problems for developers using IDEA.

Convert the annotation processors into maven plugins that execute during the `generate-sources` stage of a maven build so that IDEA can use its internal maven implementation.

**NOTE: You still need to do an initial `mvn install` once to create the `CcwCodebookVariable` enum.  Once that has been done you should be able to build and run unit tests within IDEA.**
---

### What Does This PR Do?

- Converts CodebookVariablesEnumProcessor into a maven plugin.
- Converts RifLayoutsProcessor into a maven plugin.
- Removes obsolete bfd-model-codegen-annotations module.

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:
* Verify all PR security questions and checklists have been completed and addressed.
* Does this allow you to build and run unit tests etc within IDEA.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [x] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [x] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [x] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [x] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [x] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [x] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/codebook/CodebookVariablesEnumMojo.java [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+   *     with, or <code>null</code>
+   * @param messageFormat the log message format {@link String}
+   * @param messageArguments the log message format arguments
+   */
+  private void log(String messageFormat, Object... messageArguments) {
+    String logMessage = String.format(messageFormat, messageArguments);
+    getLog().info(logMessage);
+  }
+
+  private void logError(String messageFormat, Object... messageArguments) {
```
Comment (lsmitchell) [inline apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/codebook/CodebookVariablesEnumMojo.java:191]: Missing doc

**Review by mackec1 — DISMISSED** [review]
Built branch entirely in vscode, including firing up the vscode java debugger on the project; all appears to be kosher. Lots of work went into this, thanks for tackling it.

**Review by lsmitchell — APPROVED** [review]

File: apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayoutsMojo.java [hunk 2]
```diff
@@ -64,112 +54,154 @@
...
-      }
-    } catch (RifLayoutProcessingException e) {
-      log(Diagnostic.Kind.ERROR, e.getMessage(), e.getElement());
+      outputDir = new File(outputDirectory);
+      outputDir.mkdirs();
+      process();
+      project.addCompileSourceRoot(outputDirectory);
+      // Work around bug in hibernate that causes compile to fail if meta data annotation classes
+      // already exist.
+      deleteAnnotationsDirectory();
```
[the remark this answers is not in the document]
Comment (brianburton) [reply apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayoutsMojo.java:172]: I hope not.  In that case `mvn clean` probably wouldn't work either.

File: apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayoutsMojo.java [hunk 3]
```diff
@@ -2137,4 +2083,58 @@ private static TypeName selectJavaPropertyType(
...
+   *
+   * @param messageFormat the log message format {@link String}
+   * @param messageArguments the log message format arguments
+   */
+  private void logDebug(String messageFormat, Object... messageArguments) {
+    String logMessage = String.format(messageFormat, messageArguments);
+    getLog().info(logMessage);
+  }
+
+  private void logError(String messageFormat, Object... messageArguments) {
```
[the remark this answers is not in the document]
Comment (brianburton) [reply apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayoutsMojo.java:2109]: I can add those.

**Review by KyRobbins — APPROVED** [review]

File: apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/codebook/CodebookVariablesEnumMojo.java (hunk 1 above)
Comment (brianburton) [reply apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/codebook/CodebookVariablesEnumMojo.java:191]: Comments have been added.

File: apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayoutsMojo.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (brianburton) [reply apps/bfd-model/bfd-model-codegen/src/main/java/gov/cms/bfd/model/codegen/RifLayoutsMojo.java:2144]: Comments have been added.

**Review by lsmitchell — DISMISSED** [review]

**Review by christrotteradhoc — APPROVED** [review]

**Comment by brianburton** [comment]
Closing this since the DSL maven plugin will eliminate the need for this.

**Pull request closed without merging**